README.rst
OOT library for the py.test runner
==================================

.. image:: https://pypip.in/v/pytest-oot/badge.png
    :target: https://crate.io/packages/pytest.oot/

.. image:: https://pypip.in/d/pytest-oot/badge.png
    :target: https://crate.io/packages/pytest-oot/

pytest-oot implements a simple way to write a test step for test engineers.
This plug-in use test_steps module to implement the whole things.

The test engineer can simply create test_*.oot file to use a simple case/step language,
by using the operators and options defined in test_steps module, or user-defined by you.

For detailed operators and options for steps, please refer to test_steps module:
    https://pypi.python.org/pypi?:action=display&name=test_steps


Note: while using this plug-in,
you do not need pytest-autochecklog any more. All functions have been included.


Install pytest-oot
------------------

::

    pip install pytest-oot



Example for test_*.oot file
---------------------------

Once the plug-in is installed, the pytest will automatically collect test_*.oot files
to get cases, and run each items in the files. In a test_*.oot file, each case is
a test item, and each line under it is a test step. For each test step, the syntax is
the same as a code string in checks("code string") function defined in test_steps module.



Example file: test_number.oot (you can get it from the source package)
----------------------------------------------------------------------

.. code-block:: python

    # Any words after # in a line are just comments
    # One file is a test suite. The test suite description
    test_suite: Trial1

    # Identify the test bed file, currently .py file is supported
    # similar as 'import testbed.py' in test*.py file
    # .yaml file is support too. You can use: (see more information in TestSteps package)
    # test_bed: example.test.testbed.yaml
    test_bed: example.test.testbed.py

    # A case starts from a case_idString, the description is in the bracket
    # This is to define one case, just like a function or method in a .py file
    # case_id1 means the function name is "id1"
    case: id1 (NumberBase add function):
        # under a case, there could be multiple test steps, one step in one line
        # step format: expression1 operator expression2 options
            # An expression can use any objects defined in test bed file
            # operator supports:
            #   ==(equal to), !=(not equal to), >(larger than), <(less than), >=, <=,
            #   =~(for string, contains, e.g. "hello world" =~ "llo", regex allowed
            #   !~ (not contain)
        # the following step is just like:
        #   checks( "num1.add(3,4,5,6) == 23 -t 3" )
        # in a python file, in which test_steps module is used.
        num1.add(3,4,5,6) == 23 -t 3
        num1.add(var1, var2, var3) == 18

    case: id2 (NumberBase multiple function):
        num1.multiple(2,4,5) == 200

    case: id3 (NumberChange test):
        # Every line under the case line is a step of a case
        # there could be multiple lines; each line follows the format:
        #   obj.method([parameter1 [,parameter 2 [, ...]]] operator ExpectedValue -options
        # For details, see guidance ....
        # options:
        # --timeout 30 == -t 30: fail if the step could not complete in 30 seconds
        # --repeat 30 == -r 30: repeat per second if fail until pass, timeout in 30s
        # --duration 30 == -d 30: duration of the step is 30s, if completed early, just wait until 30s
        # --expectedfail == -x true == -x: If step fail, then report pass
        # --skip == -s: just skip this step
        #
        num1.add(4)
        num2.add(3,4,5,6) == 23
        num2.multiple(4,5) == 460 -x True -t 12 -r 10
        num3.add(3,4,var2) == 1000 --skip -t 25

    case: id4 (Reverse String test):
        string1.range(1,4) == 'dlr' -d 6

    case: async1 (To test async actions - timeout)
        num_async.addw(var100, var100) == 100
        num_async.data_sync() -t 18
        num_async.get_value() == 300

    case: async2 (To test async actions - repeat)
        num_async.addw(var100, var100) >= 300
        num_async.get_value() == 500 --repeat 20

Note: the testbed file could be a python file or a .yaml file.
When it is a .py file, which defines all the variables and objects to be used in the test.
This plug-in will auto import the .py file for use in cases.
When it is a .yaml file, please follow the format definition in TestSteps package, basically,
there could be two levels of .yaml files, object file and index file.
If it reports the testbed module could not be imported, please change the module path
to make it right.


Operators & Options
-------------------

Supported Operators by default:
    == (eq), != (ne), < (lt), > (gt), <= (le), >=(ge), =~(match), !~(unmatch), =>(has), !>(hasnt)

Again, all operators defined in test_steps are supported, and you can also define them by yourself.


Supported Options by default::

    -t 30   or --timeout 30    in checks()             means       timeout=30    in check()
    -r 10   or --repeat  10    in checks()             means       repeat=10
    -d 10   or --duration 10                          means       duration=10
    -x  or --xfail or -x True or --xfail True         means       xfail=True
    -w  or --warning  or -w True  or --warning True   means       warning=True
    -s  or --skip     or -s True  or --skip True      means       skip=True
    -e MyException                                    means       exception=MyException
    -p pass_str or --passdesc pass_str                means       passdesc=pass_str
    -f fail_str or --faildesc fail_str                means       faildesc=fail_str


Test bed
--------

If you are using a test_*.oot file, you need to use

::

    testbed = [module.]testbedfilename

to import all the objects defined in the testbedfilename.py file.



Example for using step functions in a test_*.py
-----------------------------------------------

Of course, you can directly use test_steps functions in your test_*.py test scripts files.
Please refer to test_steps module for details. Some basic examples as below:


Examples (Quick Start):

1.  The Simplest step:

    .. code-block:: python

        check("num1.add(3,4,5) == 23")

    It is similar as:

    .. code-block:: python

        assert num1.add(3,4,5) == 23

    very simple, right? (we do not consider advanced features here, e.g.
    auto-logging)


2.  Step with non-python-defined operators

    .. code-block:: python

        check("string1.range(1..4) !~ r'\w\-\w'")

    Perl-like condition, =~ means 'contains', and !~ means 'not contains'.
    btw, regex can be used. The step is like:

    .. code-block:: python

        import re
        assert not re.compile(r'\w\-\w').find(string1.range(1..4))

3.  Step with timeout option

    .. code-block:: python

        check("num_async.data_sync() -t 15")

    A little complicated, -t means timeout. In this step, a time-out timer
    is set to 15 seconds. It means this step is allowed to be completed
    in 15 seconds, otherwise, it fails. no op (==, <, >, =~, etc.) in this step,
    it means no assert required to check the return value

    This is implemented by forking another thread to run the step.
    Considering some tests require to wait for a response, but how long?
    this can be useful

4.  Step with repeat option:

    .. code-block:: python

        check("num_async.get_value() == 500 --repeat 20")

    Another option --repeat (same as -r).
    The step means the step will be re-run every another second
    in total 20 seconds, until the condition comes true

    If the condition is always false in 20 seconds, then the step fails

5.  Step with multiple options

    .. code-block:: python

        check("num2.multiple(4,5) == 460 -x True -t 12 -r 10")

    Multiple options for one step ::

        -x (--expectedfail): pass if the condition is not met
        -t (--timeout): set a timeout timer
        -r (--repeat): repeat this step in 10 seconds until it comes true
           (here false actually due to -x), or timeout

6.  Use steps function to execute multiple steps

    .. code-block:: python

        checks('''
            num1.add(4)
            num2.add(3,4,5,6) == 23
            num2.multiple(4,5) == 460 -x True -t 12 -r 10
            num3.add(3,4,var2) == 1000 --skip -t 20
        ''')

7. Or you are lazy, just use s function like this:

    .. code-block:: python

       c("num2.multiple(4,5) == 460 -x True -t 12 -r 10")
       c('''
            num1.add(4)
            num2.add(3,4,5,6) == 23
            num2.multiple(4,5) == 460 -x True -t 12 -r 10
            num3.add(3,4,var2) == 1000 --skip -t 25
        ''')


Note: each line of the code strings in the checks functions can be a step in test_*.oot file.


Hooks
-----

pytest-oot is to support multiple hooks for operator, logs, and options, refer to test_steps.
Please send mails to steven004@gmail.com if you have any comments or suggestions


License
-------

This software is licensed under the `MIT license <http://en.wikipedia.org/wiki/MIT_License>`_.

© 2015 Steven LI



CHANGES.rst
Changelog
=========

0.6.2
-----

- Support case: (new format for case, and easy for editing in an IDE)
    and case_ (to back-compatible to previous versions)

0.6.1
-----

- Support .yaml test bed file for initialization of all objects defined in a test bed.
- An example is provided in the example/test folder.
- See more in the TestSteps package

0.5.6
-----

- Fix the log-new-func issue.

0.5.0
-----

- Big change, use test_steps module to simplify the implementation, oot_step.py is omitted
- More support of options and operators added.
- Format is not so strict any more. Each step just follow the format in step function defined in test_steps


0.3.0
-----

- Implement the pytest standard test result outcome for test_*.oot file test.
- Hide trace-back to the step functions, so that the result is clearer


0.2.7
-----

- Update the license file and PYPI related documentation.


0.2.4
-----

- The first version of the implementation of pytest_oot plugin and step functions



example/test/testbed.py
__author__ = 'Steven LI'

from ..my_calculate import *
from ..my_string import *

num1 = Num_Base(5)
num2 = Num_Change(5)
num3 = Num_Change(10)
num4 = Num_Change(5)
num_async = Num_Asyc(100)

string1 = Reverse("Hello World!")

var1 = 3
var2 = 4
var3 = 5

var100 = 100

example/test/test_string.py
__author__ = 'Steven LI'

from .testbed import *

def test_string1():
    ret = string1.range(1,4)
    assert ret == 'dlr'


example/test/test_example.py
__author__ = 'Steven LI'

from .testbed import *

def test_NumBase_add():
    ret = num1.add(3,4,5,6)
    assert ret == 23

def test_NumBase_mul():
    ret = num1.multiple(2,4,5)
    assert ret == 100, "result is not 100"

def test_NumChange():
    ret = num2.add(3,4,5,6)
    assert ret == 23
    ret = num2.multiple(4,5)
    assert ret == 460

def test_NumChange3():
    ret = num3.add(3,4,5,6)
    assert ret == 100


example/my_calculate.py
__author__ = 'Steven LI'
import time

class Num_Base:
    def __init__(self, base=0):
        self.data = base

    def add(self, *args):
        ret_val = self.data
        for val in args:
            ret_val += val
        return ret_val

    def multiple(self, *args):
        ret_val = self.data
        for val in args:
            ret_val *= val
        return ret_val

class Num_Change:
    def __init__(self, base=0):
        self.data = base

    def add(self, *args):
        for val in args:
            self.data += val
        return self.data

    def multiple(self, *args):
        for val in args:
            self.data *= val
        return self.data

import random
class Num_Asyc(Num_Base):
    def addw(self, *args):
        self.data2 = self.data
        self.start = time.time()
        self.wait = True
        self.sleep = random.uniform(1, 20)
        print(self.sleep)
        for val in args:
            self.data2 += val
        #print(self.data2)
        return self.data

    def mulw(self, *args):
        self.data2 = self.data
        self.start = time.time()
        self.wait = True
        for val in args:
            self.data2 *= val
        #print(self.data2)
        return self.data

    def get_value(self):
        if self.wait:
            self.end = time.time()
            if self.end - self.start > self.sleep:
                self.data = self.data2
                self.wait = False

        return self.data

    def data_sync(self):
        if not self.wait: return
        while(True):
            self.end = time.time()
            if self.end - self.start > self.sleep:
                self.data = self.data2
                self.wait = False
                return
            time.sleep(1)

import sys
import threading
class KThread2(threading.Thread):
    """A subclass of threading.Thread, with a kill()
    method.

    Come from:
    Kill a thread in Python:
    http://mail.python.org/pipermail/python-list/2004-May/260937.html
    """

    def __init__(self, *args, **kwargs):
        threading.Thread.__init__(self, *args, **kwargs)
        self.killed = False

    def start(self):
        """Start the thread."""
        self.__run_backup = self.run
        self.run = self.__run      # Force the Thread to install our trace.
        threading.Thread.start(self)

    def __run(self):
        """Hacked run function, which installs the
        trace."""
        sys.settrace(self.globaltrace)
        self.__run_backup()
        self.run = self.__run_backup

    def globaltrace(self, frame, why, arg):
        if why == 'call':
          return self.localtrace
        else:
          return None

    def localtrace(self, frame, why, arg):
        if self.killed:
          if why == 'line':
            raise SystemExit()
        return self.localtrace

    def kill(self):
        self.killed = True

if __name__ == '__main__':
    num_1 = Num_Asyc(5)
    num_1.addw(3,4,5)

    null_list = ()
    num_1.data_sync(*null_list)
    print(num_1.get_value(*null_list))

    import threading
    #t = KThread2(target=num_1.data_sync)
    t = threading.Thread(target=num_1.data_sync)
    t.start()
    t.join(7)
    if t.is_alive():
        #t.kill()
        raise TimeoutError()
    print(num_1.get_value())


pytest_oot/oot_plugin.py
"""
    pytest_oot
    ~~~~~~~~~~~~~~~

    Helpers for defining a simple format for object-oriented system-level testing
    The design is implemented taking pytest_yamlwsgi as a reference.

    :copyright: 2014-2015 Steven LI <steven004@gmail.com>
    :license: MIT

    This module is presented as a py.test plugin. It provides two main features:

        1. Provide a method to easy write test scripts in a plain format, easy to review and work, and
        2. Provide some options for test steps to simplify the test scripts writing. It could make 10 lines python code in one line

    How it works
    ------------

    py.test has a number of hooks that can be accessed to extend its
    behaviour. This plugin uses the `pytest_collect_file` hook which is called
    wih every file available in the test target. The first action of the
    hook-handler is that it checks if the filename is of the form `test*.oot`,
    here oot means object-oriented test since in the script file, every step is
    based on an object to take an action and compare the result.

    If it finds an appropriate file, the hook-handler will generate a number
    of tests based on that file. py.test provides a hierarchical system of
    tests. The top-level of this hierarchy is an object representing the oot
    file itself. The lower components are represented by further test
    containers, and subsequent individual tests.

    The test*.oot file is just like a test*.py file, but in different format, Let's
    see an example first. (you can get the code from the package source)
        # Any words after # in a line are just comments
        test_suite: Trial1
        # Identify the test bed file, currently .py file is supported
        # similar as 'import testbed.py' in test*.py file
        test_bed: testbed

        # A case starts from a case_idString, the description is in the bracket
        # This is to define one case, just like a function or method in a .py file
        # case_id1 means the function name is "id1"
        case_id1 (NumberBase add function):
            # under a case, there could be multiple test steps, one step in one line
            # step format: obj.method(parameters) operator expected_result options
                # obj/methods are defined in test bed file
                # operator supports:
                #   ==(equal to), !=(not equal to), >(larger than), <(less than), >=, <=,
                #   =~(for string, contains, e.g. "hello world" =~ "llo", regex allowed
                #   !~ (not contain)
            num1.add(3,4,5,6) == 23 -t 3
            num1.add(var1, var2, var3) == 18

        case_id2 (NumberBase multiple function):
            num1.multiple(2,4,5) == 100

        case_id3 (NumberChange test):
            # Every line under the case line is a step of a case
            # there could be multiple lines; each line follows the format:
            #   obj.method([parameter1 [,parameter 2 [, ...]]] operator ExpectedValue -options
            # For details, see guidance ....
            # options:
            # --timeout 30 == -t 30: fail if the step could not complete in 30 seconds
            # --repeat 30 == -r 30: repeat per second if fail until pass, timeout in 30s
            # --duration 30 == -d 30: duration of the step is 30s, if completed early, just wait until 30s
            # --expectedfail == -x true == -x: If step fail, then report pass
            # --skip == -s: just skip this step
            #
            num1.add(4)
            num2.add(3,4,5,6) == 478
            num2.multiple(4,5) == 460 -x True -t 12 -r 10
            num3.add(3,4,var2) == 1000 --skip -t 25

        case_id4 (Reverse String test):
            string1.range(1,4) == 'dlr' -d 6

        case_async1 (To test async actions - timeout)
            num_async.addw(var100, var100) == 100
            num_async.data_sync() -t 15
            num_async.get_value() == 300

        case_async2 (To test async actions - repeat)
            num_async.addw(var100, var100) >= 300
            num_async.get_value() == 500 --repeat 20


    Each case of the test suite should be a subclass of:class:`py.test.collect.Item`,
    which is a direct subclass, for leaf tests, implementing `runtest()`.

"""

__author__ = 'Steven LI'

import test_steps
import pytest
import re


def pytest_configure(config):
    test_steps.auto_func_detection(False)


def pytest_runtest_setup(item):
    test_steps.log_new_func(item.name, str(item.fspath))


def pytest_collect_file(parent, path):
    if path.ext == ".oot" and path.basename.startswith("test"):
        return TestCaseFile(path, parent)


class TestCaseFile(pytest.File):
    # test_bed: the test bed file
    # test_cases: an array of the test cases

    def collect(self):
        suite_content = self.fspath.open().read()
        self.__parse_suite(suite_content)

        # Import objects in the test bed
        if self.test_bed:
            # self.objs = __import__(self.test_bed, globals())
            # import importlib
            # self.objs = importlib.import_module(self.test_bed)
            self.objs = test_steps.init_testbed(self.test_bed)

        current_line_number = 0
        case_number = 0
        for case in self.cases:
            case_id = case[0:case.find(' ')].lstrip()
            current_line_number += self.case_lines[case_number]
            case_number += 1
            yield TestCaseItem(case_id, self, case, current_line_number)

    def __parse_suite(self, suite_content):
        ''' To parse the system case file, see detail of the system test file example
        :param suite_content: The content of the test case file.
        :return: no return, exception will be raised if the format is not right
        '''

        # split cases, notice that the first element of the cases is about the test suite description
        # cases = suite_content.split("\ncase_")
        cases = re.compile('\n[cC]ase[_:]').split(suite_content)
        self.case_lines = [c.count('\n') + 1 for c in cases]

        # Deal with the test suite description; cases[0] is about the test suite summary
        self.name = 'unknown'
        self.test_bed = None
        self.suite_attr = {}
        self.cases = cases[1:]

        # cases[0] is not a case, actually it's the suite header
        header = cases[0].split("\n")
        for line in header:
            line = line.strip()
            if len(line) == 0: continue
            if line[0] == '#': continue
            colon = line.find(":")
            if colon == -1:
                continue   # no ':', it is not a key-value pair, ignore it
            (name, spec) = (line[0:colon], line[colon + 1:].strip())
            if name == 'test_suite':
                self.name = spec
            elif name == 'test_bed':
                self.test_bed = spec
            else:
                self.suite_attr[name] = spec


class TestCaseItem(pytest.Item):
    ''' A case structure, one case can contain multiple steps
    '''

    # def __init__(self, name, parent, case_dec, steps):
    #     super(TestCaseItem, self).__init__(name, parent)
    #     self.case_dec = case_dec
    #     self.steps = steps

    def __init__(self, case_id, parent, case_string, line_number):
        super(TestCaseItem, self).__init__(case_id, parent)
        header_end = case_string.find('\n')
        self.case_header = case_string[0:header_end].strip()
        self.steps = case_string[header_end:].split('\n')
        self.first_line = line_number
        self.parent = parent
        # m = re.match(r'(\w+)\s*\((.*)\)', line1)
        # (case_id, case_dec) = m.group(1,2)

    def runtest(self):
        self.current_step = 0
        for step_string in self.steps:
            # step_string = step_string.strip()
            # if len(step_string) == 0: continue
            # if step_string[0] == '#': continue
            # step_obj = TestStep(step, self) # To create a step_obj with parsing
            # step_obj.execute()
            test_steps.steps(step_string, self.parent.objs.__dict__)

            self.current_step += 1

    def repr_failure(self, excinfo):
        """ called when self.runtest() raises an exception. """

        failed_line_number = self.first_line + self.current_step + 1

        fail_string = "case_" + self.case_header + "\n".join(self.steps[0:self.current_step])
        cur_step_str = self.steps[self.current_step]
        #       fail_string += '\n>' + self.steps[self.current_step][1:]
        fail_string += '\n>' + cur_step_str[1:]
        fail_string += "\nE" + ' ' * (len(cur_step_str) - len(cur_step_str.lstrip()) - 1) + excinfo.value.args[1]

        fail_string += "\n\n%s:" % self.parent.fspath + "%d" % failed_line_number

        # if isinstance(excinfo.value, TestStepFail):
        #     fail_string += ": TestStepFail"
        #
        # elif isinstance(excinfo.value, TestRunTimeError):
        #     fail_string += ": TestRunTimeError"
        fail_string += ": " + excinfo.value.__class__.__name__

        return fail_string

    def reportinfo(self):
        ''' Called when there is a failure as a failed case title'''
        return self.fspath, 0, "%s" % self.case_header


pytest_oot/oot_step.py
"""
This is to implement a function called step(), and a bunch-function steps() or s():
The format of a step looks like:
    obj.method(parameter) op exp-value options
In this one step, there is an action, and also a check point
This one step can be translated to multiple lines of python code,
or dozens lines of code if there is one option or multiple options.

How to import:
from pytest_oot import step, steps, s

Examples (Quick Start):
1.  step("num1.add(3,4,5) == 23")
    the same as:
        assert num1.add(3,4,5) == 23
very simple, right?

2.  step("string1.range(1..4) !~ r'\w\-\w'")
    Perl-like condition, =~ means 'contains', and !~ means 'not contains'
            btw, regex can be used
    The step is like:
        import re
        assert re.compile(r'\w\-\w').find(string1.range(1..4))

3.  step("num_async.data_sync() -t 15")
    a little complicated, -t means timeout. In this step, a time-out timer is set to 15 seconds.
    It means this step is allowed to be completed in 15 seconds, otherwise, it fails.
    no op (==, <, >, =~, etc.) in this step, it means no assert required to check the return value

    This is implemented by forking another thread to run the step.
    Considering some tests require to wait for a response, but how long? this can be useful

4.  step("num_async.get_value() == 500 --repeat 20")
    Another option --repeat (same as -r).
    The step means the step will be re-run every another second in total 20 seconds,
        until the condition comes true
    If the condition is always false in 20 seconds, then the step fails

5.  step("num2.multiple(4,5) == 460 -x True -t 12 -r 10")
    Multiple options for one step.
        -x (--expectedfail): pass if the condition is not met
        -t (--timeout): set a timeout timer
        -r (--repeat): repeat this step in 10 seconds until it comes true(here false actually due to -x), or timeout
6.  steps('''
        num1.add(4)
        num2.add(3,4,5,6) == 23
        num2.multiple(4,5) == 460 -x True -t 12 -r 10
        num3.add(3,4,var2) == 1000 --skip -t 25
    ''')
    multiple steps in one shot
7. Or you are lazy, just use s function like this:
   s("num2.multiple(4,5) == 460 -x True -t 12 -r 10")
   s('''
        num1.add(4)
        num2.add(3,4,5,6) == 23
        num2.multiple(4,5) == 460 -x True -t 12 -r 10
        num3.add(3,4,var2) == 1000 --skip -t 25
    ''')
"""


__author__ = 'Steven LI'

import time
import threading
import re
from inspect import getmodule, currentframe
#import pytest

def step(step_string, test_module=None, parent=None):
    __tracebackhide__ = True
    step = step_string.strip()
    if len(step) == 0: return
    if step[0] == '#': return

    if not test_module:
        test_module = getmodule(currentframe().f_back)
    test_step = TestStep(step_string, test_module, parent)
    test_step.parse()
    test_step.execute()

def steps(step_strings):
    __tracebackhide__ = True
    step_list = step_strings.split('\n')
    test_module = getmodule(currentframe().f_back)
    for step_string in step_list:
        step(step_string, test_module)

# a short-style of steps function
s = steps

class TestStep:
    '''A test step class, to deal with all test step related actions'''
    def __init__(self, step_string, test_module, parent=None):
        self.__default_set__()

        self.step_namespace = test_module.__dict__

        self.step_string_o = step_string
        self.step_string = step_string.strip()
        self.parent = parent
        #self.parse()

    def __default_set__(self):
        #self.obj   # the obj string
        #self.method    # the method string
        self.func = None  # the actual function object to be invoked (not the string)
        self.params = ()    # the list of parameters
        self.op = None
        self.exp_value = None
        self.options = {
            # 0: no timeout setting; others, raise Timeout exception if no return in Timeout seconds
            # --timeout   == -t
            "Timeout": 0,

            # 0: no duration setting; Stay on this step for xxx seconds if not 0
            # --duration == -d
            "Duration": 0,

            # Ture: this step is expected to fail
            # --expectedfailure  == -x
            "ExpectedFailure": False,

            # Repeat this step until expected value returned, the number setting means timeout seconds
            # --repeat  == -r
            "RepeatInSeconds": 0,

            # To skip this step
            # --skip  == -s
            "Skip": False
        }

    def execute(self):
        __tracebackhide__ = True
        if self.options["Skip"]: return
        end_time = time.time() + self.options["Duration"]
        if self.options["Timeout"] > 0 :
            t = threading.Thread(target = self.__execute)
            t.start()
            t.join(self.options["Timeout"])
            if t.is_alive():
                raise TestStepFail(self.step_string, self.__result_exp__("Timeout"))
        else:
            self.__execute()

        ## Deal with the Duration Options
        wait_seconds = end_time - time.time()
        if wait_seconds >0: time.sleep(wait_seconds)


    def __execute(self):
        __tracebackhide__ = True
        if self.options["RepeatInSeconds"] > 0:
            end_time = time.time() + self.options["RepeatInSeconds"]
            while (time.time() < end_time) :
                self.ret = self.func(*self.params)
                if self.__step_pass() :
                    return
                time.sleep(1)
            raise TestStepFail(self.step_string, self.__result_exp__())
            #pytest.fail("really failed. .... ", True)
        else :
            self.ret = self.func(*self.params)
            if not self.__step_pass():
                raise TestStepFail(self.step_string, self.__result_exp__())
                #pytest.fail("failed ........", True)

    def __result_exp__(self, msg = None):
        m = re.compile(r'(\s*)\w.*').match(self.step_string_o)
        if msg:
            return m.group(1) + msg
        else:
            return m.group(1) + '%r %s %s' %(self.ret, self.op, self.exp_value)

    def __step_pass(self):  # deal with the ExpectedFailure option
        if not self.op: return True
        if self.options["ExpectedFailure"] :
            return (not self.step_pass())
        else :
            return self.step_pass()

    def step_pass(self):
        ret_val = self.ret
        if self.exp_value == None: #There is an operator, but no expected value
            raise TestRunTimeError(self.step_string, self.__result_exp__("Expected_value required"))

        if isinstance(ret_val, str):
            if self.op in ['=~', '!=']:
                l = re.compile(self.exp_value).find(ret_val)
                return True if (self.op == '=~' and l) or (self.op == '!~' and not l) else False
            else: # for ==, !=, <, >, <= and >= of strings
                return eval('ret_val' + self.op + self.exp_value)
        elif isinstance(ret_val, float):
            import math
            try:
                exp_value_f = float(self.exp_value)            # The expected result is a number?
            except ValueError:
                raise TestRunTimeError(self.step_string, self.__result_exp__("expected value is expected to be a float"))
            if self.op in ['<', '<=', '>', '>=']:
                return eval('ret_val' + self.op + 'exp_value_f')
            elif self.op == '==': # to see if they are very close
                return math.fabs(ret_val - exp_value_f) < 1.0e-6  # precise 1.0e-6
            elif self.op == '!=' :
                return math.fabs(ret_val - exp_value_f) > 1.0e-6  # precise 1.0e-6
            else:
                raise TestRunTimeError(self.step_string, self.__result_exp__("Wrong operator for a float"))
        elif isinstance(ret_val, int):
            if not re.compile(r"\d*(\.\d+)?").match(self.exp_value):
                raise TestRunTimeError(self.step_string, self.__result_exp__("expected value is expected to be a integer"))
            if self.op in ['<', '<=', '>', '>=', '==', '!=']:
                return eval('ret_val' + self.op + self.exp_value)
            else:
                raise TestRunTimeError(self.step_string, self.__result_exp__("Wrong operator for a float"))


    def parse(self):
        __tracebackhide__ = True
        regex = re.compile(r'(\w+)\.(\w+)\s*(\(.*\))(.*)$')
        m = regex.match(self.step_string)
        (self.obj, self.method, param_str, p_f) = m.group(1,2,3,4)
        #obj = getattr(self.objs, self.obj)
        obj = self.step_namespace[self.obj]
        self.func = getattr(obj, self.method)
        try: #if wrong, just keep null list
            self.params = eval(param_str[0:-1] + ",)", self.step_namespace, {})  # the list of parameters
        except:
            self.params = ()
        p_f = p_f.strip()

        if len(p_f) != 0 and p_f[0] != '#':
            try:
                m = re.compile(r'(==|!=|<=|>=|<|>|=~|!~)\s*(\"[^\"]+\"|\'[^\']+\'|\S+)(.*)$').match(p_f)
                (self.op, self.exp_value, options) = m.group(1,2,3)
                options = options.strip()
            except:
                options = p_f

            if len(options) != 0 and options[0] != '#':
                ol = re.compile(r'(?<!^)\s+(?=(?:-\w|--\w{2,}))').split(options)
                for o in ol:
                    if o[0] != '-' :
                        raise TestRunTimeError(self.step_string, self.__result_exp__("wrong options: "+o))
                    if o[1] == '-': # It is a long option string
                        r = re.compile(r'--(\w+)\s*(\S*)$').match(o)
                        if r.group(1) == "timeout":
                            self.options['Timeout'] = int(r.group(2))
                        elif r.group(1) == "duration":
                            self.options['Duration'] = int(r.group(2))
                        elif r.group(1) == 'expectedfailure':
                            if len(r.group(2))==0 or r.group(2)[0] in "TtYy":
                                self.options['ExpectedFailure'] = True
                        elif r.group(1) == 'skip':
                            if len(r.group(2))==0 or r.group(2)[0] in "TtYy":
                                self.options['Skip'] = True
                        elif r.group(1) == 'repeat' :
                            self.options['RepeatInSeconds'] = int(r.group(2))
                        else: raise TestRunTimeError(self.step_string, self.__result_exp__("wrong options: "+o))
                    else:
                        key = o[1]
                        if len(o) > 2: value = o[2:].strip()
                        else: value = ''
                        if key == "t":
                            self.options['Timeout'] = int(value)
                        elif key == "d":
                            self.options['Duration'] = int(value)
                        elif key == 'x':
                            if len(value)==0 or value[0] in "TtYy":
                                self.options['ExpectedFailure'] = True
                        elif key == 's':
                            if len(value)==0 or value[0] in "TtYy":
                                self.options['Skip'] = True
                        elif key == 'r' :
                            self.options['RepeatInSeconds'] = int(value)
                        else: raise TestRunTimeError(self.step_string, self.__result_exp__("wrong options: "+o))


class TestStepFail(Exception):
    """ custom exception for error reporting. """

class TestRunTimeError(Exception):
    """ custom exception for error reporting. """


